### DISCUSSION TITLE
Can't import modules in next.config

### DISCUSSION THREAD

**Discussion by florianwalther-private** [vercel/next.js#39130] [Help]
How can I import a module in `next.config`? I always get "module not found" even tho the path is correct.

This is how my code looks like (simplified):

import testFile from './testfile';

```
// @ts-check

/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  images: {
    domains: ["server.tuthub.io", "localhost", "www.tuthub.io"]
  },
  async redirects() {
    [...]
  },
  async rewrites() {
    [...] // In here I'm trying to use testFile
  }
}
```

**Comment by GabenGar** [comment]
Are you running it in [ESM mode](https://nodejs.org/api/esm.html)?

**Reply by florianwalther-private** [reply]
I tried out both, it works with neither setting.

**Reply by GabenGar** [reply]
How do you export your module in the `./testfile`? Also ESM module imports have to include the file extension.

**Reply by florianwalther-private** [reply]
> How do you export your module in the `./testfile`? Also ESM module imports have to include the file extension.

I have deleted the file meanwhile but I tried out every different combination. I have the suspicion that it's just not possible. Here is an unsolved Stackoverflow with the same problem:https://stackoverflow.com/questions/70381402/nextjs-next-config-mjs-import-local-constant-file

**Reply by GabenGar** [reply]
Well it does have the same issue of not providing the file extension in the import path.

**Reply by florianwalther-private** [reply]
I tried it, but I get "An import path cannot end with a '.ts' extension. Consider importing '../filename.js' instead."

**Reply by GabenGar** [reply]
You can't import typescript files into the config because it isn't a part of typescript build. Either rewrite it into vanilla js or add a separate typescript build step for that file.

**Reply by florianwalther-private** [reply]
@GabenGar I'll try it out and report back. I can't change that file right now.

**Comment by john-pimq** [comment]
I also face the same issues.

`theme.js`
```
export default {
  primaryColor: 'red',
}
```

`next.config.mjs`
```
// Not work
import theme from './theme' ;

// It need to add the file extension and work
import theme from './theme.js' ;
```

------

Change the file extension can solve this problem. However if my `theme.js` import another files, it can not easily be resolved.

`theme.js`
```
import btnTheme from './btn';
import colorTheme from './color';
// and more style setting files...

export default {
  ...btnTheme,
  ...colorTheme
}
```

`next.config.mjs`
```
import theme from './theme.js' ;
```

and it would throw the error `Error [ERR_MODULE_NOT_FOUND]: Cannot find module '[filepath]/btn'`.

------

The problem is show in the version
```
"next": "^12.3.2"
```

**Reply by GabenGar** [reply]
Why are you having your config as a ESM module while the rest of the code as CJS?

**Reply by john-pimq** [reply]
I think the rest of my code is write in ESM (not CJS). It use `import` and `export default`.

**Reply by GabenGar** [reply]
No, it's written in ESM-transpiled-to-CJS-before-node-ESM-existed.
Extensions are required in Node ESM module imports and your setup doesn't make sense in the first place. Just put `"type": "module"` in your `package.json` and then fix all bugs until the project builds. Then all your `.js` files will be assumed as ESM modules and you don't need to deal with importing ESM modules into CJS modules.

**Reply by collinsedim** [reply]
Renaming `theme.js` to `theme.mjs` should fix the problem. You can then import as `import theme from './theme.mjs' ` . It works fine for me.

**Comment by OscarJVD** [comment]
same problem
